docs/howtos/full_eval.rst

```diff
@@ -0,0 +1,161 @@
+
+.. code-block:: python
+
+  collections.Counter(
+      tuple(batch['image'].shape)
+      for batch in tfds.load('mnist', split='test').batch(per_device_batch_size)
+  )
+  # output:
+  # Counter({(272, 28, 28, 1): 1, (512, 28, 28, 1): 19})
+
+The problem is accentuated when using multiple devices for data parallelism.  If
+the batch size is not **divisible by the number devices**, then that last step
+must be executed on a single device (or a subset of devices). Usually one would
+drop the last batch, but this will lead to incorrect results.
+
+
+.. code-block:: python
+
+  sum(
+      np.prod(batch['label'].shape)
+      for batch in tfds.load('mnist', split='test')
+          .batch(per_device_batch_size, drop_remainder=True)
+          .batch(jax.local_device_count())
+  )
+  # output:
+  # 9728
+
+Using multiple hosts further complicates the situation because JAX uses the SPMD
+paradigm and every host must execute the same program. We would usually form
+non-overlapping splits for different hosts with |tfds.split_for_jax_process()|_,
+but this can lead to **different numbers for different hosts**, resulting in
+different JAX programs when all examples are to be processed.
+
+.. code-block:: python
+
+  process_count = 6
+  [
+      len(tfds.load(dataset_name, split=tfds.split_for_jax_process(
+          'test', process_index=process_index, process_count=process_count)))
+      for process_index in range(process_count)
+  ]
+  # output:
+  # [1667, 1667, 1667, 1667, 1666, 1666]
+
+
+
+.. |tfds.split_for_jax_process()| replace:: ``tfds.split_for_jax_process()``
+.. _tfds.split_for_jax_process(): https://www.tensorflow.org/datasets/api_docs/python/tfds/split_for_jax_process
+
+
+The solution: padding
+---------------------
+
+Even though it's possible to solve this problem by cleverly adjusting the number
+of batches executed by different devices on different hosts, such a solution
+quickly becomes complicated and makes the main eval loop hard to read with a lot
+of cumbersome logic.
+
+The more straight forward solution to this problem is to use padding at the end
+of the dataset to make sure that the last batch has the same size as the
+preceding batches.
+
+
+Manual implementation
+~~~~~~~~~~~~~~~~~~~~~
+
+The last batch is manually padded to contain the same number of examples as in
+the preceding batches. The predictions for the padded examples are discarded
+from the computation.
+
+.. code-block:: python
+
+  shard = lambda x: einops.rearrange(
+      x, '(d b) ... -> d b ...', d=jax.local_device_count())
+  unshard = lambda x: einops.rearrange(x, 'd b ... -> (d b) ...')
+
+  correct = total = 0
+  for batch in ds.as_numpy_iterator():
+    images = batch['image']
+    n = len(images)
+    padding = np.zeros([per_host_batch_size - n, *images.shape[1:]], images.dtype)
+    padded_images = np.concatenate([images, padding])
+    preds = unshard(get_preds(vs_p, shard(padded_images)))[:n]
```

Comment:
Hey! What does `vs_p` mean / represent? Is it short for "variables_predict"? If it is, it took me a while to figure out, maybe a more descriptive name would help readers 😀

Reply:
I abbreviated `variables` with `vs`. Then often when I replicate things I use the `_p` suffix. But I'm not sure if that use is really widespread.

In either case, the `_p` is not needed here since it's an empty set, so I removed it, and replaced `vs` with the full `variables`.

Reply:
> In either case, the _p is not needed here since it's an empty set, so I removed it, and replaced `vs` with the full `variables`.

I still see `vs_p` here, maybe I am wrong but is the intention is to rename it to `variables`?